docs/map.rst

```diff
@@ -0,0 +1,602 @@
+   a runtime error if the map is empty.
+
+Modification
+^^^^^^^^^^^^
+
+Adding a new entry to a map
+"""""""""""""""""""""""""""
+
+::
+
+    Map.insert :: Ord k => k -> v -> Map k v -> Map k v
+    Map.insert key value m = ...
+
+:haddock_short:`Data.Map.Strict#insert` adds the ``value`` into the map ``m``
+with the given ``key``, replacing the existing value if the key already exists.
+
+::
+
+    Map.insert 1 "one" Map.empty
+    > Map.fromList [(1,"one")]
+
+    Map.insert 4 "four" (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > fromList [(1,"one"),(2,"two"),(3,"three"),(4,"four")]
+
+    Map.insert 1 "uno" (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > fromList [(1,"uno"),(2,"two"),(3,"three")]
+
+
+Removing an entry from a map
+""""""""""""""""""""""""""""
+
+::
+
+    Map.delete :: Ord k => k -> Map k v -> Map k v
+    Map.delete key m = ...
+
+:haddock_short:`Data.Map.Strict#delete` removes the entry with the specified
+``key`` from the map ``m``.  If the key doesn't exist it leaves the map
+unchanged. Remember, maps are immutable so if you delete an entry from a map you
+need to assign the new map to a new variable.
+
+::
+
+    Map.delete 1 Map.empty
+    > Map.empty
+
+    Map.delete 1 (Map.fromList [(1,"one"),(2,"two"),(3,"three")])
+    > fromList [(2,"two"),(3,"three")]
+
+Filtering map entries
+"""""""""""""""""""""
+
+::
+
+    Map.filterWithKey :: (k -> v -> Bool) -> Map k v -> Map k v
+    Map.filterWithKey predicate m = ...
+
+:haddock_short:`Data.Map.Strict#filterWithKey` produces a map consisting of all
+entries of ``m`` for which the ``predicate`` returns ``True``.
+
+::
+
+    let f key value = key == 2 || value == "one"
+    Map.filterWithKey f (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > fromList [(1,"one"),(2,"two"]
+
+
+Modifying a map entry
+"""""""""""""""""""""
+
+::
+
+    Map.adjust :: Ord k => (v -> v) -> k -> Map k v -> Map k v
+    Map.adjust f key m = ...
+
+:haddock_short:`Data.Map.Strict#abjust` applies the value transformation
+function ``f`` to the entry with given ``key``. If no entry for that key exists
+then the map is left unchanged.
+
+::
+
+    Map.alter :: Ord k => (Maybe v -> Maybe v) -> k -> Map k v -> Map k v
+    Map.alter f key m = ...
+
+Apply the value transformation function ``f`` to the entry with given ``key``,
+if no entry for that key exists then the function is passed ``Nothing``. If the
+function returns ``Nothing`` then the entry is deleted, if the function returns
+``Just v2`` then the value for the ``key`` is updated to ``v2``. In other words,
+alter can be used to insert, update, or delete a value.
+
+::
+
+    let removeElement _ = Nothing
+    Map.alter removeElement "key" (Map.fromList [("key", 0)])
+    > fromList []
+
+    let setValueToOne _ = Just 1
+    Map.alter setValueToOne "key" (Map.fromList [("key", 0)])
+    > fromList [("key",1)]
+
+    Map.alter setValueToOne "key" Map.empty
+    > fromList [("key",1)]
```

Comment:
Neither of these examples actually explain what really makes `alter` useful &mdash;you've just written likely-slower versions of `insert` and `delete`! You could also rewrite `adjust`, but there's not much point. `alter` gets useful if

* You don't know whether you want to change (or retain) or delete a key until you've seen what value it's mapped to, and/or

* You only want to insert a value if it's not already present.

Reply:
Good call, I just copied these from the Haddocks, which means we should make the haddock examples better as well :) I'm trying to make everything in the examples copy/pasteable into a ghci session which doesn't work too well with case statements. How would you feel about something like this:

```
let addValueIfMissing mv = if isJust mv then mv else (Just 1)
```

An example where we want to look at the value would require a case statement, I think I'll just make a separate block and mention that it must be in a .hs source file.

```
doubleIfPositive :: Maybe Int -> Maybe Int
doubleIfPositive mv = case mv of
  Just v -> if v > 0 then (Just v*2) else (Just v)
  Nothing -> Just 1

Map.alter doubleIfPositive "key" (Map.fromList [("key", 1)])
```

Thoughts?

Reply:
Updated the examples.